src/config.cpp

```diff
@@ -201,6 +202,172 @@ std::string toAbsolute(const fs::path& base, const fs::path& path) {
+                                            debugStr);
+    const auto parseCommon = [&](auto& input) {
+        input.module = module;
+        parseMandatory(valueIt, "input_type", debugStr, input.inputType);
+        parseMandatory(valueIt, "delay", debugStr, input.delay);
+        parseMandatory(valueIt, "duration", debugStr, input.duration);
+        parseMandatory(valueIt, "node_set", debugStr, input.nodeSet);
+    };
+
+    switch (module) {
+    case Module::linear: {
+        SimulationConfig::InputLinear ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "amp_start", debugStr, ret.ampStart);
+        parseOptional<double>(valueIt, "amp_end", ret.ampEnd, ret.ampStart);
+        return ret;
+    }
+    case Module::relative_linear: {
+        SimulationConfig::InputRelativeLinear ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "percent_start", debugStr, ret.percentStart);
+        parseOptional<double>(valueIt, "percent_end", ret.percentEnd, ret.percentStart);
+        return ret;
+    }
+    case Module::pulse: {
+        SimulationConfig::InputPulse ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "amp_start", debugStr, ret.ampStart);
+        parseMandatory(valueIt, "width", debugStr, ret.width);
+        parseMandatory(valueIt, "frequency", debugStr, ret.frequency);
+        parseOptional<double>(valueIt, "amp_end", ret.ampEnd, ret.ampStart);
+        return ret;
+    }
+    case Module::subthreshold: {
+        SimulationConfig::InputSubthreshold ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "percent_less", debugStr, ret.percentLess);
+        return ret;
+    }
+    case Module::noise: {
+        SimulationConfig::InputNoise ret;
+        parseCommon(ret);
+        const auto mean = valueIt.find("mean");
+        const auto mean_percent = valueIt.find("mean_percent");
+
+        if (mean != valueIt.end()) {
+            parseOptional(valueIt, "mean", ret.mean);
+        }
+
+        if (mean_percent != valueIt.end()) {
+            parseOptional(valueIt, "mean_percent", ret.meanPercent);
+        }
+
+        if (ret.mean.has_value() && ret.meanPercent.has_value()) {
+            throw SonataError(
+                fmt::format("Both `mean` or `mean_percent` have values in {}", debugStr));
+        } else if (!ret.mean.has_value() && !ret.meanPercent.has_value()) {
+            throw SonataError(
+                fmt::format("One of `mean` or `mean_percent` need to have a value in {}",
+                            debugStr));
+        }
+
+        parseOptional(valueIt, "variance", ret.variance);
+        return ret;
+    }
+    case Module::shot_noise: {
+        SimulationConfig::InputShotNoise ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "rise_time", debugStr, ret.riseTime);
+        parseMandatory(valueIt, "decay_time", debugStr, ret.decayTime);
+        parseOptional<int>(valueIt, "random_seed", ret.randomSeed, randomSeed);
```

Comment:
I remember that the template argument deduction should be working here (like in the `report` section), so no need to specify the types.

Reply:
sure, fixed